sympy/parsing/mathematica.py

```diff
@@ -588,7 +590,21 @@ def parse(self, s):
         "##": lambda: ["SlotSequence", "1"],
     }
 
-    _literal = r"[A-Za-z][A-Za-z0-9]*"
+    # This regex matches any valid python identifier -- excluding
+    # underscores, which Mathematica uses to denote patterns, and
+    # therefore can't be part of a variable name.  The regex has the
+    # form "[a][b]*", where `a` is the set of characters that can
+    # start an identifier, and `b` is the set of characters that can
+    # continue an identifier, which may also include numbers and
+    # unicode combining characters.
+    _literal = (
+        "["
+        + "".join(c for c in map(chr, range(sys.maxunicode+1)) if c!="_" and c.isidentifier())
+        + "]["
+        + "".join(c for c in map(chr, range(sys.maxunicode+1)) if c!="_" and ("x"+c).isidentifier())
+        + "]*"
+    )
```

Comment:
This is creating a megabyte sized regex.

Is it actually reasonable to use a regex for something like this?

Would it not make more sense to just use the `isidentifier` method somewhere else?

Reply:
The problem is that this regex goes into a larger regex that is then used to separate a string into tokens.  So I don't know a reasonably easy way to find a substring that *could* be a token and then test it with `isidentifier` without using a big regex.  In any case, that's the existing approach that I piggybacked on.

I will say that this is created and finds tokens very quickly for me, and won't be created unless someone actually is parsing Mathematica.  I.e., it won't be an extra MB hanging around for everyone using sympy.

That said, this isn't the smartest way to create this sub-regex; it could use ranges instead.  A quick experiment suggests it could be ~75x smaller.  I'll try to get around to that in a few days.